### PR TITLE
Add StartupQueue automatically; update tests.

### PR DESCRIPTION
**Pull request by bsneed**
- Adds StartupQueue to the initial list of plugins.
- Adds waitUntilStarted(analytics:) to test suite.
- Modifies some tests to wait until started.

### CODE REVIEW

**Review by prayansh — APPROVED** [review]

File: Tests/Segment-Tests/Support/TestUtilities.swift [hunk 1]
```diff
@@ -107,3 +107,16 @@ class OutputReaderPlugin: Plugin {
...
+func waitUntilStarted(analytics: Analytics?) {
+    guard let analytics = analytics else { return }
+    var started = false
+    while started == false {
+        if let system: System = analytics.store.currentState() {
+            if system.started {
+                started = true
+            }
+        }
+        RunLoop.main.run(until: Date.distantPast)
```
Comment (migs647) [inline Tests/Segment-Tests/Support/TestUtilities.swift:120]: Would like a comment here saying you're ending the current runloop incase some newbs don't recognize it and take it out.

File: Tests/Segment-Tests/Support/TestUtilities.swift [hunk 2]
```diff
@@ -107,3 +107,16 @@ class OutputReaderPlugin: Plugin {
         return event
     }
 }
+
+func waitUntilStarted(analytics: Analytics?) {
+    guard let analytics = analytics else { return }
+    var started = false
+    while started == false {
+        if let system: System = analytics.store.currentState() {
```
Comment (migs647) [inline Tests/Segment-Tests/Support/TestUtilities.swift:115]: Wouldn't mind some commented context =). Makes sense right now in the moment, but what about three months from now?

File: Tests/Segment-Tests/Support/TestUtilities.swift (hunk 1 above)
Comment (migs647) [reply Tests/Segment-Tests/Support/TestUtilities.swift:120]: nevermind, just noticed this is in tests

File: Tests/Segment-Tests/Support/TestUtilities.swift (hunk 2 above)
Comment (migs647) [reply Tests/Segment-Tests/Support/TestUtilities.swift:115]: nevermind, just noticed this is in tests

**Review by migs647 — APPROVED** [review]
LGTM

File: Sources/Segment/Utilities/QueueTimer.swift [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            return
+        }
+        state = .resumed
+        timer.resume()
+    }
+}
+
+
+extension TimeInterval {
+    static func milligarvins(_ value: Int) -> TimeInterval {
```
Comment (prayansh) [inline Sources/Segment/Utilities/QueueTimer.swift]: i think u got a typo here 😂

Comment (migs647) [reply Sources/Segment/Utilities/QueueTimer.swift]: 🤦‍♀️

File: Sources/Segment/Settings.swift [hunk 4]
```diff
@@ -52,18 +52,49 @@ public struct Settings: Codable {
...
+        return l == r
+    }
+}
+
 extension Analytics {
-    func checkSettings() {
+    internal func update(settings: Settings) {
+        apply { (plugin) in
+            // tell all top level plugins to update.
+            update(plugin: plugin, settings: settings)
```
Comment (migs647) [inline Sources/Segment/Settings.swift:67]: Do we want to maybe discriminate on only destination plugins here or do we want to track "started" on all types of plugins? I'm cool with it being all, just want to make sure we're on the same page.

File: Sources/Segment/Settings.swift [hunk 5]
```diff
@@ -79,10 +79,23 @@ extension Analytics {
     }
     
     internal func checkSettings() {
+        if isUnitTesting {
+            // we don't really wanna wait for this network call during tests...
+            // but we should make it work similarly.
+            store.dispatch(action: System.ToggleRunningAction(running: false))
+            DispatchQueue.main.async {
```
Comment (migs647) [inline Sources/Segment/Settings.swift:86]: Do we want `[weak self] in` for the closure?

File: Sources/Segment/Startup.swift [hunk 6]
```diff
@@ -66,34 +66,46 @@ extension Analytics: Subscriber {
...
 extension Analytics {
     internal func setupSettingsCheck() {
+        // do the first one
+        checkSettings()
+        // set up return-from-background to do it again.
         NotificationCenter.default.addObserver(forName: UIApplication.willEnterForegroundNotification, object: nil, queue: OperationQueue.main) { (notification) in
-            self.checkSettings()
+            guard let app = notification.object as? UIApplication else { return }
+            if app.applicationState == .background {
+                self.checkSettings()
```
Comment (migs647) [inline Sources/Segment/Startup.swift:79]: I'm still unsettled about using lifecycle events for triggering this.

File: Sources/Segment/Startup.swift [hunk 7]
```diff
@@ -66,34 +66,46 @@ extension Analytics: Subscriber {
...
 import Cocoa
 extension Analytics {
     internal func setupSettingsCheck() {
-        NotificationCenter.default.addObserver(forName: NSApplication.willBecomeActiveNotification, object: nil, queue: OperationQueue.main) { (notification) in
+        // do the first one
+        checkSettings()
+        // now set up a timer to do it every 24 hrs.
+        // mac apps change focus a lot more than iOS apps, so this
+        // seems more appropriate here.
+        QueueTimer.schedule(interval: 5, queue: .main) {
```
Comment (migs647) [inline Sources/Segment/Startup.swift]: This is so much better than using a selector 👍

File: Sources/Segment/Utilities/QueueTimer.swift [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        return TimeInterval(other: ((value / 60) * 1000) / 42)
```
Comment (migs647) [inline Sources/Segment/Utilities/QueueTimer.swift]: Funny guy... but what is the 42 for?

**Review by migs647 — APPROVED** [review]
LGTM

File: Sources/Segment/Settings.swift (hunk 4 above)
Comment (bsneed) [reply Sources/Segment/Settings.swift:67]: yeah, should be all.

File: Sources/Segment/Settings.swift (hunk 5 above)
Comment (bsneed) [reply Sources/Segment/Settings.swift:86]: yes, good call out.

**Pull request merged**